File: backend/app/services/workspace_service.py
import json
import re
import shutil
from datetime import UTC, datetime
from pathlib import Path

from fastapi import HTTPException

from app.config import settings
from app.schemas.workspace import Workspace, WorkspaceCreate

WORKSPACE_SUBDIRS = ("notes", "drawings", "resources", "assets/images")
CONFIG_FILENAME = "config.json"
_SLUG_PATTERN = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")


def _slugify(name: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", name.strip().lower()).strip("-")
    return slug or "workspace"


def _unique_slug(base_slug: str) -> str:
    slug = base_slug
    counter = 2
    while (settings.workspaces_root / slug).exists():
        slug = f"{base_slug}-{counter}"
        counter += 1
    return slug


def get_workspace_dir(workspace_id: str) -> Path:
    # workspace_id comes straight from the URL path, and gets joined onto a
    # real filesystem path below — reject anything that isn't a plain slug
    # before it can be used to escape workspaces_root (e.g. "../../etc").
    if not _SLUG_PATTERN.match(workspace_id):
        raise HTTPException(status_code=404, detail="Workspace not found")

    workspace_dir = settings.workspaces_root / workspace_id
    if not workspace_dir.is_dir() or not (workspace_dir / CONFIG_FILENAME).exists():
        raise HTTPException(status_code=404, detail="Workspace not found")

    return workspace_dir


def _read_config(workspace_dir: Path) -> dict:
    return json.loads((workspace_dir / CONFIG_FILENAME).read_text())


def _write_config(workspace_dir: Path, config: dict) -> None:
    (workspace_dir / CONFIG_FILENAME).write_text(json.dumps(config, indent=2))


def list_workspaces() -> list[Workspace]:
    root = settings.workspaces_root
    if not root.exists():
        return []

    workspaces = []
    for entry in sorted(root.iterdir()):
        config_path = entry / CONFIG_FILENAME
        if not entry.is_dir() or not config_path.exists():
            continue
        try:
            config = _read_config(entry)
            workspaces.append(
                Workspace(id=entry.name, name=config["name"], created_at=config["created_at"])
            )
        except (json.JSONDecodeError, KeyError):
            continue

    return workspaces


def create_workspace(data: WorkspaceCreate) -> Workspace:
    name = data.name.strip()
    if not name:
        raise HTTPException(status_code=422, detail="Workspace name cannot be empty")

    slug = _unique_slug(_slugify(name))
    workspace_dir = settings.workspaces_root / slug

    for subdir in WORKSPACE_SUBDIRS:
        (workspace_dir / subdir).mkdir(parents=True, exist_ok=True)

    created_at = datetime.now(UTC)
    _write_config(workspace_dir, {"name": name, "created_at": created_at.isoformat()})

    return Workspace(id=slug, name=name, created_at=created_at)


def rename_workspace(workspace_id: str, name: str) -> Workspace:
    name = name.strip()
    if not name:
        raise HTTPException(status_code=422, detail="Workspace name cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    config = _read_config(workspace_dir)
    config["name"] = name
    _write_config(workspace_dir, config)

    return Workspace(id=workspace_id, name=name, created_at=config["created_at"])


def delete_workspace(workspace_id: str) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    shutil.rmtree(workspace_dir)

File: backend/src/services/workspaceService.ts
import fs from 'fs';
import path from 'path';
import { settings } from '../config.js';
import { Workspace, WorkspaceCreate } from '../types/workspace.js';

const WORKSPACE_SUBDIRS = ['notes', 'drawings', 'resources', 'assets/images'];
const CONFIG_FILENAME = 'config.json';
const SLUG_PATTERN = /^[a-z0-9]+(-[a-z0-9]+)*$/;

function slugify(name: string): string {
  const slug = name
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
  return slug || 'workspace';
}

function uniqueSlug(baseSlug: string): string {
  let slug = baseSlug;
  let counter = 2;
  while (fs.existsSync(path.join(settings.workspacesRoot, slug))) {
    slug = `${baseSlug}-${counter}`;
    counter++;
  }
  return slug;
}

export function getWorkspaceDir(workspaceId: string): string {
  if (!SLUG_PATTERN.test(workspaceId)) {
    throw { status: 404, message: 'Workspace not found' };
  }

  const workspaceDir = path.join(settings.workspacesRoot, workspaceId);
  if (!fs.existsSync(workspaceDir) || !fs.statSync(workspaceDir).isDirectory() || !fs.existsSync(path.join(workspaceDir, CONFIG_FILENAME))) {
    throw { status: 404, message: 'Workspace not found' };
  }

  return workspaceDir;
}

function readConfig(workspaceDir: string): any {
  return JSON.parse(fs.readFileSync(path.join(workspaceDir, CONFIG_FILENAME), 'utf8'));
}

function writeConfig(workspaceDir: string, config: any): void {
  fs.writeFileSync(path.join(workspaceDir, CONFIG_FILENAME), JSON.stringify(config, null, 2), 'utf8');
}

export function listWorkspaces(): Workspace[] {
  const root = settings.workspacesRoot;
  if (!fs.existsSync(root)) return [];

  const entries = fs.readdirSync(root);
  const workspaces: Workspace[] = [];

  for (const entry of entries) {
    const entryDir = path.join(root, entry);
    if (!fs.statSync(entryDir).isDirectory()) continue;

    const configPath = path.join(entryDir, CONFIG_FILENAME);
    if (!fs.existsSync(configPath)) continue;

    try {
      const config = readConfig(entryDir);
      workspaces.push({
        id: entry,
        name: config.name,
        created_at: config.created_at
      });
    } catch (e) {
      continue;
    }
  }

  // Sort workspaces alphabetically by ID/slug name matching Python sorted()
  return workspaces.sort((a, b) => a.id.localeCompare(b.id));
}

export function createWorkspace(data: WorkspaceCreate): Workspace {
  const name = data.name.trim();
  if (!name) {
    throw { status: 422, message: 'Workspace name cannot be empty' };
  }

  const slug = uniqueSlug(slugify(name));
  const workspaceDir = path.join(settings.workspacesRoot, slug);

  for (const subdir of WORKSPACE_SUBDIRS) {
    fs.mkdirSync(path.join(workspaceDir, subdir), { recursive: true });
  }

  const createdAt = new Date().toISOString();
  writeConfig(workspaceDir, { name, created_at: createdAt });

  return { id: slug, name, created_at: createdAt };
}

export function renameWorkspace(workspaceId: string, name: string): Workspace {
  const trimmedName = name.trim();
  if (!trimmedName) {
    throw { status: 422, message: 'Workspace name cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const config = readConfig(workspaceDir);
  config.name = trimmedName;
  writeConfig(workspaceDir, config);

  return { id: workspaceId, name: trimmedName, created_at: config.created_at };
}

export function deleteWorkspace(workspaceId: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  fs.rmSync(workspaceDir, { recursive: true, force: true });
}
